### DISCUSSION TITLE
Slow response as many routes exist

### DISCUSSION THREAD

**Discussion by shadow3x3x3** [apache/apisix#7720] [Q&A]
All config of apisix is default.

There are 20,000 routes in apisix. We added a new route and found that responding to the new route takes about 1.8 seconds for an HTTP response. Is this normal?

Thank you.

**Comment by tzssangglass** [comment]
Not normal, we can add the prometheus plugin to see where the delay is?

**Reply by shadow3x3x3** [reply]
Thanks for replying! After we updated the etcd version from v3.4 to v3.5, responses from 20,000 routes are pretty good.

But when we increased the routes number to 200,000 routes, there were too many 502 errors from apisix responses. How can we tune apisix to support large numbers of routes? or to use more powerful machines?

We didn't enable any plugins and every route only includes a single upstream.

**Reply by tzssangglass** [reply]
502 seems to be an upstream problem that APISIX cannot connect to upstream. Any logs about 502?

**Reply by shadow3x3x3** [reply]
Sorry, but the response was 504 Gateway-Timeout after we checked the message, not 502.

**Reply by tokers** [reply]
@shadow3x3x3 Could you give some details about the `504`? Also, error logs will be useful.

**Reply by shadow3x3x3** [reply]
We found that if there are many routes in apisix (about 200,000) then some responses would be extremely slow (curl took 326s) and the CPU usage of apisix was also high (almost 100%). The response content is OK without any error log.

The test request parameters are the same, but response times are different.

The only one line in `error.log`:

```log
2022/08/18 03:27:25 [warn] 29513#29513: *4701856 upstream timed out {110: Connection timed out} while connecting to upstream, client: 127.0.0.1, server: _, request: "GET / HTTP/1.1", upstream: "http:/1.1.1.1:80/", host: "test8199.com:9080"
```

**Reply by tzssangglass** [reply]
`upstream: "http:/1.1.1.1:80/` was it an error or was that the original log?
This error should be related to the `1.1.1.1` is not a real upstream. Or APISIX cannot connect to this upstream, which is a problem with the network environment.

BTW, 200,000 routes, are you testing or do you really need 200,000 routes?

**Reply by tokers** [reply]
@shadow3x3x3 What's your scenario?

**Reply by shadow3x3x3** [reply]
@tzssangglass Yes, that's the original log, but that's okay. I think the error was caused by heavy loading during testing.

We want to use APISIX for our multiple services. If APISIX can hold 200,000 routes, it might be a better choice than other api gateways for us.

**Reply by tokers** [reply]
@shadow3x3x3 So you're testing the extreme case?

**Reply by shadow3x3x3** [reply]
@tokers Yes.

**Reply by tzssangglass** [reply]
> Yes, that's the original log, but that's okay. I think the error was caused by heavy loading during testing.

I think the 502 issue is caused by bad upstream format and address(`http:/1.1.1.1:80/`) than heavy loading ,you could change the upstream to a normally accessible address to rule out if the 502 issue is arising from this.

**Reply by moonming** [reply]
@shadow3x3x3 This is an interesting scenario, we can talk in more detail, I am the PMC chair of Apache APISIX. 
You can reach me at wenming@apache.org or https://calendly.com/wenming/30min
